src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -382,6 +382,13 @@ namespace service_nodes
                       continue;
                     }
 
+                    if (height - *cryptonote::get_hard_fork_heights(m_core.get_nettype(), hf_version).first < netconf.HARDFORK_DEREGISTRATION_GRACE_PERIOD) {
+                      LOG_PRINT_L2("Decommissioned service node "
+                                   << quorum->workers[node_index]
+                                   << " is still not passing required checks, and has no remaining credits left. However it is within the grace period of a hardfork so has not been deregistered.");
+                      continue;
+                    }
```

Comment:
There's also a case where we vote to deregister below this (i.e. if the SN has no credit and so goes directly from active -> dereg).  Rather than duplicating it there, I think you can just move this check down into a `if (vote_for_state == new_state::deregister && height - *....) { /* log and continue */ }` just before the `service_nodes::make_state_change_vote(...)` call.

Reply:
Should delete this chunk now?

Reply:
ugh, think i brought this back when rebasing